src/main/scala/com/databricks/spark/csv/package.scala

```diff
@@ -87,36 +89,41 @@ package object csv {
         case None => None
       }
 
+      val csvFormatBase = CSVFormat.DEFAULT
+        .withDelimiter(delimiterChar)
+        .withEscape(escapeChar)
+        .withSkipHeaderRecord(false)
+        .withNullString("null")
+
+      val csvFormat = quoteChar match {
+        case Some(c) => csvFormatBase.withQuote(c)
+        case _ => csvFormatBase
+      }
+
       val generateHeader = parameters.getOrElse("header", "false").toBoolean
       val header = if (generateHeader) {
-        dataFrame.columns.map(c => s""""$c"""").mkString(delimiter)
+        csvFormat.format(dataFrame.columns.map(_.asInstanceOf[AnyRef]):_*)
       } else {
         "" // There is no need to generate header in this case
       }
       val strRDD = dataFrame.rdd.mapPartitions { iter =>
-        val csvFormatBase = CSVFormat.DEFAULT
-          .withDelimiter(delimiterChar)
-          .withEscape(escapeChar)
-          .withSkipHeaderRecord(false)
-          .withNullString("null")
-
-        val csvFormat = quoteChar match {
-          case Some(c) => csvFormatBase.withQuote(c)
-          case _ => csvFormatBase
-        }
-
         new Iterator[String] {
           var firstRow: Boolean = generateHeader
 
-          override def hasNext = iter.hasNext
+          override def hasNext = iter.hasNext || firstRow
```

Comment:
I see why you added it. Would you please leave a comment as to why. Future contributors will have an easier time understanding it.